### ISSUE TITLE
Deterministic nonce on ECDSA signing

### ISSUE THREAD

**Issue by OnedgeLee** [planetarium/libplanet#2240]
While do ECDSA signing with `Libplanet.Crypto.PrivateKey`, it seems like nonce `k` is deterministic with respect to `privateKey` and `message`.

Method [`Libplanet.Crypto.PrivateKey.Sign`](https://github.com/planetarium/libplanet/blob/494a9ed20190d5a33477882b79a17922e45aee7e/Libplanet/Crypto/PrivateKey.cs#L220-L224) calls [`Libplanet.Crypto.DefaultCryptoBackend.Sign`](https://github.com/planetarium/libplanet/blob/494a9ed20190d5a33477882b79a17922e45aee7e/Libplanet/Crypto/DefaultCryptoBackend.cs#L13-L35).
```c#
// Libplanet.Crypto.PrivateKey.cs
public ImmutableArray<byte> Sign(ImmutableArray<byte> message)
{
    HashDigest<SHA256> hashed = HashDigest<SHA256>.DeriveFrom(message);
    byte[] sig = CryptoConfig.CryptoBackend.Sign(hashed, this);
    return Unsafe.As<byte[], ImmutableArray<byte>>(ref sig);
}
```

[`Libplanet.Crypto.DefaultCryptoBackend.Sign`](https://github.com/planetarium/libplanet/blob/494a9ed20190d5a33477882b79a17922e45aee7e/Libplanet/Crypto/DefaultCryptoBackend.cs#L13-L35) utilizes nonce generator on [`Org.BouncyCastle.Crypto.Signers.HMacDsaKCalculator`](https://github.com/chrishaly/bc-csharp/blob/022e86493b52c8c8dfe414940acc3a04cdfa3382/crypto/src/crypto/signers/HMacDsaKCalculator.cs#L35-L136), and it only supports deterministic nonce generation.
```c#
// Libplanet.Crypto.DefaultCryptoBackend.cs
public byte[] Sign(HashDigest<T> messageHash, PrivateKey privateKey)
{
    var h = new Sha256Digest();
    var kCalculator = new HMacDsaKCalculator(h);
    var signer = new ECDsaSigner(kCalculator);
    signer.Init(true, privateKey.KeyParam);
    BigInteger[] rs = signer.GenerateSignature(messageHash.ToByteArray());
    var r = rs[0];
    var s = rs[1];

    BigInteger otherS = privateKey.KeyParam.Parameters.N.Subtract(s);
    if (s.CompareTo(otherS) == 1)
    {
        s = otherS;
    }

    var bos = new MemoryStream(72);
    var seq = new DerSequenceGenerator(bos);
    seq.AddObject(new DerInteger(r));
    seq.AddObject(new DerInteger(s));
    seq.Close();
    return bos.ToArray();
}
```

```c#
// Org.BouncyCastle.Crypto.Signers.HMacDsaKCalculator.cs
public virtual bool IsDeterministic => true;
public virtual void Init(BigInteger n, SecureRandom random)
{
    throw new InvalidOperationException("Operation not supported");
}
```
So it seems like there are no options for stochastic nonce generation on [`Org.BouncyCastle.Crypto.Signers.ECDsaSigner.GenerateSignature`](https://github.com/chrishaly/bc-csharp/blob/022e86493b52c8c8dfe414940acc3a04cdfa3382/crypto/src/crypto/signers/DsaSigner.cs#L82-L107), which is signing algorithm for [`Libplanet.Crypto.DefaultCryptoBackend.Sign`](https://github.com/planetarium/libplanet/blob/494a9ed20190d5a33477882b79a17922e45aee7e/Libplanet/Crypto/DefaultCryptoBackend.cs#L13-L35).
```c#
// Org.BouncyCastle.Crypto.Signers.ECDsaSigner.cs
public virtual BigInteger[] GenerateSignature(byte[] message)
{
    ECDomainParameters parameters = key.Parameters;
    BigInteger n = parameters.N;
    BigInteger bigInteger = CalculateE(n, message);
    BigInteger d = ((ECPrivateKeyParameters)key).D;
    if (kCalculator.IsDeterministic)
    {
        kCalculator.Init(n, d, message);
    }
    else
    {
        kCalculator.Init(n, random);
    }

    ECMultiplier eCMultiplier = CreateBasePointMultiplier();
    BigInteger bigInteger3;
    BigInteger bigInteger4;
    while (true)
    {
        BigInteger bigInteger2 = kCalculator.NextK();
        bigInteger3 = eCMultiplier.Multiply(parameters.G, bigInteger2).Normalize().AffineXCoord.ToBigInteger().Mod(n);
        if (bigInteger3.SignValue != 0)
        {
            bigInteger4 = bigInteger2.ModInverse(n).Multiply(bigInteger.Add(d.Multiply(bigInteger3))).Mod(n);
            if (bigInteger4.SignValue != 0)
            {
                break;
            }
         }
     }

     return new BigInteger[2] { bigInteger3, bigInteger4 };
}
```

Current `k` generation logic on [`Org.BouncyCastle.Crypto.Signers.HMacDsaKCalculator.Init(BigInteger n, BigInteger d, byte[] message)`](https://github.com/chrishaly/bc-csharp/blob/022e86493b52c8c8dfe414940acc3a04cdfa3382/crypto/src/crypto/signers/HMacDsaKCalculator.cs#L45-L97) and  [`Org.BouncyCastle.Crypto.Signers.HMacDsaKCalculator.NextK()`](https://github.com/chrishaly/bc-csharp/blob/022e86493b52c8c8dfe414940acc3a04cdfa3382/crypto/src/crypto/signers/HMacDsaKCalculator.cs#L99-L136) seems like to use [`Deterministic DSA and ECDSA` on RFC6979](https://www.rfc-editor.org/rfc/rfc6979.html#section-3.2).

Since it follows the process on [RFC6979](https://www.rfc-editor.org/rfc/rfc6979.html#section-3.2), and it won't use same `k` for different messages, I think it won't be generate potential security vulnerabilities(safe from private key reconstruction under nonce reuse), but have no idea if which one is better, stochastic or deterministic nonce generator.

May be we can discuss about it, and provide both of nonce generation methods.

**Comment by longfin** [comment]
The point I've remembered totally wrong is, libsecp256k1 was already adopted RFC 6979 since https://github.com/bitcoin-core/secp256k1/pull/163. (I don't know why I remembered wrong about determisitism of secp256k1 😅 )

[PS3 private key leakage](https://en.wikipedia.org/wiki/PlayStation_3_homebrew#Private_key_compromised) is a representative anecdote in which we feel that it is dangerous to use a deterministic nonce. however, on the contrary, it can also be interpreted as an anecdote that it is more dangerous to leave it to the user to implement non-deterministic randomness, and it seems safer to deterministically calculate it from other entropy (i.e., private key) as suggested in RFC 6979.

So I think it's okay not to change the current deterministic signature structure.
